feat: add updateMetadata func to token cache

bshaffer opened the pull request:
This function, while not officially implemented as part of `FetchAuthTokenInterface`, is defined on all Credentials instances, and _should_ be a part of `FetchAuthTokenCache` as well.

This is required so the cache can take advantage of the logic in `updateMetadata` for defaulting to self-signed JWTs (see googleapis/gax-php#272)

dwsupplee commented on src/FetchAuthTokenCache.php:
```diff
@@ -188,4 +188,27 @@ public function getProjectId(callable $httpHandler = null)
+     * @param callable $httpHandler callback which delivers psr7 request
+     * @return array updated metadata hashmap
+     */
+    public function updateMetadata(
```
Should this end up calling the underlying implementation from the class it wraps? It looks like the implementations vary between credentials type.

bshaffer replied:
This is a great point. The reason I didn't do it this way is because I thought this would bypass the cache, as calling `parent::updateMetadata` would result in `parent::fetchAuthToken` method being called instead of `FetchAuthTokenCache::fetchAuthToken`. However, I think now that `parent::updateMetadata` would execute`FetchAuthTokenCache::fetchAccessToken` instead, and so making this a wrapper call is the way to go.
 We just need to ensure a test confirms this.

bshaffer replied:
After further consideration, this doesn't work because the Cache is a wrapper and not a subclass.

So we can:

1. Bypass the logic in `updateMetadata` - this is bad for obvious reasons
2. Bypass the cache in `fetchAuthToken` - this renders the class useless if `updateMetadata` is used
3. Cache the result of `fetchAuthToken` separately - this is somewhat useful, but still not optimal
4. Use the passed-in `$metadata` array to include the cached value, and update the existing classes so they only fill in the header if it isn't set - I don't think we want this behavior.
5. Leave `updateMetadata` out of the picture until v2 - this is the best option I see for now.

bshaffer commented:
After *even further* consideration I have done the following

 - Added logic to bypass fetching a token if the `authorization` key is set in the metadata. This allows the cached access token to propagate w/o the fetcher fetching a new one
 - Added `UpdateMetadataInterface` for better implementation checking
 - Added `UpdateMetadataTrait` for dryer/cleaner code

Also important:
 - moved the `AUTH_METADATA_KEY` constant from `CredentialsLoader` into new `UpdateMetadataInterface`
 - Implemented `UpdateMetadataInterface` in `GCECredentials`, `FetchAuthTokenCache` and `CredentialsLoader`
 - used `UpdateMetadataTrait` in `GCECredentials` and `CredentialsLoader`
 - Added tests for the new functionality!

dwsupplee commented on src/UpdateMetadataInterface.php:
```diff
@@ -0,0 +1,41 @@
+ * Describes a Credentials object which supports updating request metadata
+ * (request headers).
+ */
+interface UpdateMetadataInterface
```
Should the `IAMCredentials` class implement this too? The auth keys seem to differ in that implementation a bit though, which makes it seem like the constant being on this class might not be ideal.

bshaffer replied:
I don't really know how the `IAMCredentials` class is being used, so I left it out for now. It doesn't even extend `FetchAuthTokenInterface` so it makes more sense for me to leave it out.

dwsupplee replied:
Do we need the interface then? Perhaps I missed it but it looks like anything that extends `CredentialsLoader` will implement the `updateMetadata` method. If the goal is to provide better implementation checking, we can do that by checking if the credentials are an instance of `CredentialsLoader` I believe.

dwsupplee replied:
After thinking on this a bit, the interface is better, let's keep it.

bshaffer replied:
The interface is used by `FetchAuthTokenCache`, which does not extend `CredentialsLoader`. It's only real function is to ensure that the `updateMetadata` function exists on the fetcher, since we can't add that method to `FetchAuthTokenInterface` without breaking backwards compatibility.

It isn't strictly necessary but I think it's best practice to have it. Not only is the interface cleaner, but it ensures that the function signature is compatible.

bshaffer commented on src/Credentials/GCECredentials.php:
```diff
@@ -58,7 +59,8 @@
     ProjectIdProviderInterface,
-    GetQuotaProjectInterface
+    GetQuotaProjectInterface,
+    UpdateMetadataInterface
```
For some reason I thought this didn't extend credentials loader... So you're right, this can be removed

bshaffer replied:
Removed!

bshaffer commented on src/CredentialsLoader.php:
```diff
@@ -26,13 +26,16 @@
     const NON_WINDOWS_WELL_KNOWN_PATH_BASE = '.config';
-    const AUTH_METADATA_KEY = 'authorization';
+
+    use UpdateMetadataTrait;
```
See above... I thought we needed the logic here in two places, but I was mistaken. I can move it back into credentials loader...

bshaffer replied:
Removed `UpdateMetadataTrait`!

bshaffer commented:
Any thoughts on this @dwsupplee or @jdpedrie ? I'm hoping to get this sucker into `v1.12.0`

jdpedrie reviewed (approved):
LGTM
